module/applications/journal/class-summary-sheet.mjs

```diff
@@ -0,0 +1,439 @@
+    if ( hasFeatures ) headers[0].push({content: game.i18n.localize("DND5E.Features")});
+    headers[0].push(...scaleValues.map(a => ({content: a.title})));
+    if ( spellProgression ) {
+      if ( spellProgression.headers.length > 1 ) {
+        headers[0].forEach(h => h.rowSpan = 2);
+        headers[0].push(...spellProgression.headers[0]);
+        headers[1] = spellProgression.headers[1];
+      } else {
+        headers[0].push(...spellProgression.headers[0]);
+      }
+    }
+
+    const cols = [{ class: "level", span: 1 }];
+    if ( item.type === "class" ) cols.push({class: "prof", span: 1});
+    if ( hasFeatures ) cols.push({class: "features", span: 1});
+    if ( scaleValues.length ) cols.push({class: "scale", span: scaleValues.length});
+    if ( spellProgression ) cols.push(...spellProgression.cols);
+
+    const makeLink = uuid => TextEditor._createContentLink(["", "UUID", uuid]).outerHTML;
+
+    const rows = [];
+    for ( const level of Array.fromRange((CONFIG.DND5E.maxLevel - (initialLevel - 1)), initialLevel) ) {
+      const features = [];
+      for ( const advancement of item.advancement.byLevel[level] ) {
+        switch ( advancement.constructor.typeName ) {
+          case "ItemGrant":
+            if ( advancement.configuration.optional ) continue;
+            features.push(...advancement.configuration.items.map(makeLink));
+            continue;
+        }
+      }
+
+      // Level & proficiency bonus
+      const cells = [{class: "level", content: level.ordinalString()}];
+      if ( item.type === "class" ) cells.push({class: "prof", content: `+${Math.floor((level + 7) / 4)}`});
+      if ( hasFeatures ) cells.push({class: "features", content: features.join(", ")});
+      scaleValues.forEach(s => cells.push({class: "scale", content: s.formatValue(level)}));
+      const spellCells = spellProgression?.rows[rows.length];
+      if ( spellCells ) cells.push(...spellCells);
+
+      // Skip empty rows on subclasses
+      if ( (item.type === "subclass") && !features.length && !scaleValues.length && !spellCells ) continue;
+
+      rows.push(cells);
+    }
+
+    return { headers, cols, rows };
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Build out the spell progression data.
+   * @param {Item5e} item  Class item belonging to this journal.
+   */
+  async _getSpellProgression(item) {
+    const spellcasting = item.spellcasting;
+    if ( !spellcasting || (spellcasting.progression === "none") ) return null;
+
+    const items = [item.toObject()];
+    let classId = item.id;
+    if ( item.type === "subclass" ) {
+      classId = foundry.utils.randomID();
+      items.push({_id: classId, name: "tmp-class", type: "class", system: {identifier: item.system.classIdentifier}});
+    }
+    const actor = await Actor.implementation.create({name: "tmp", type: "character", items}, {temporary: true});
```

Comment:
It's a clever way to reuse our spell preparation code, but it does a feel a little hacky to me. Can we instead factor out the spell preparation code in such a way that it will accept, say, a list of class items and return a structure that's suitable for use here as well as part of actor preparation?

Reply:
I have another branch where I am reworking spell preparation, so that can be handled there rather than adding more to this PR.

Reply:
Is that a PR already? Or just a branch? What's the scope of the refactor there?

Reply:
Just a branch for now: https://github.com/arbron/dnd5e/blob/spell/api/module/documents/actor/actor.mjs#L608

This splits `_prepareSpellcasting` into several separate methods and cleans up the logic a bunch, as well as adding hooks to allow for overriding default spellcasting preparation or adding custom types.

Reply:
It doesn't look too big or controversial. Can you submit it as a PR maybe and I can prioritise it. I'd rather we do it that way than have this code merged as-is, in case it does get merged and then we never get around to the spellcasting PR in this release.

Reply:
Okay, PR created: #1888